src/runtime_src/core/edge/drm/zocl/edge/zocl_hwctx.c

```diff
@@ -0,0 +1,460 @@
+
+    mutex_lock(&client->lock);
+
+    kds_hw_ctx = kds_get_hw_ctx_by_id(client, drm_cu_ctx->hw_context);
+    if (!kds_hw_ctx) {
+        DRM_ERROR("%s: No valid hw context is open", __func__);
+        ret = -EINVAL;
+        goto out;
+    }
+
+    ret = zocl_cu_ctx_to_info(zdev, drm_cu_ctx, kds_hw_ctx, &kds_cu_info);
+    if (ret) {
+        DRM_ERROR("%s: No valid CU context found for this hw context", __func__);
+        goto out;
+    }
+
+    kds_cu_ctx = kds_alloc_cu_hw_ctx(client, kds_hw_ctx, &kds_cu_info);
+    if (ret) {
+        DRM_ERROR("%s: Allocation of CU context failed", __func__);
+        ret = -EINVAL;
+        goto out;
+    }
+
+    ret = kds_add_context(&zdev->kds, client, kds_cu_ctx);
+    if (ret) {
+        DRM_ERROR("%s: Failed to add kds context", __func__);
+        kds_free_cu_ctx(client, kds_cu_ctx);
+        goto out;
+    }
+
+    drm_cu_ctx->cu_index = set_domain(kds_cu_ctx->cu_domain, kds_cu_ctx->cu_idx);
+
+out:
+    mutex_unlock(&client->lock);
+    return ret;
+}
+
+int zocl_close_cu_ctx(struct drm_zocl_dev *zdev, struct drm_zocl_close_cu_ctx *drm_cu_ctx, struct kds_client *client)
+{
+    struct kds_client_hw_ctx *kds_hw_ctx = NULL;
+    struct kds_client_cu_ctx *kds_cu_ctx = NULL;
+    struct kds_client_cu_info kds_cu_info = {};
+    int ret = 0;
+
+    mutex_lock(&client->lock);
+
+    kds_hw_ctx = kds_get_hw_ctx_by_id(client, drm_cu_ctx->hw_context);
+    if (!kds_hw_ctx) {
+        DRM_ERROR("%s: No valid hw context is open", __func__);
+        ret = -EINVAL;
+        goto out;
+    }
+
+    zocl_close_cu_ctx_to_info(drm_cu_ctx, &kds_cu_info);
+
+    kds_cu_ctx = kds_get_cu_hw_ctx(client, kds_hw_ctx, &kds_cu_info);
+    if (!kds_cu_ctx) {
+        DRM_ERROR("%s: No cu context is open", __func__);
+        ret = -EINVAL;
+        goto out;
+    }
+
+    ret = kds_del_context(&zdev->kds, client, kds_cu_ctx);
+    if (ret)
+        goto out;
+
+    ret = kds_free_cu_ctx(client, kds_cu_ctx);
+
+out:
+    mutex_unlock(&client->lock);
+    return ret;
+}
+
+/**
+ * Callback function for async dma operation. This will also clean the
+ * command memory.
+ *
+ * @arg:    kds command pointer
+ * @ret:    return value of the dma operation.
+ */
+static void zocl_kds_dma_complete(void *arg, int ret)
+{
+    struct kds_command *xcmd = (struct kds_command *)arg;
+    zocl_dma_handle_t *dma_handle = (zocl_dma_handle_t *)xcmd->priv;
+
+    xcmd->status = KDS_COMPLETED;
+    if (ret)
+        xcmd->status = KDS_ERROR;
+    xcmd->cb.notify_host(xcmd, xcmd->status);
+    xcmd->cb.free(xcmd);
+
+    kfree(dma_handle);
+}
+
+/**
+ * Copy the user space command to kds command. Also register the callback
+ * function for the DMA operation.
+ *
+ * @zdev:   zocl device structure
+ * @flip:   DRM file private data
+ * @ecmd:   ERT command structure
+ * @xcmd:   KDS command structure
+ *
+ * @return      0 on success, Error code on failure.
+ */
+static int copybo_ecmd2xcmd(struct drm_zocl_dev *zdev, struct drm_file *filp,
```

Comment:
Please give a different name. Even it's a static but same name is already exists which lead a confusion later. Better to give name as copybo_hwxtc_ecmd2xcmd.

Same applicable to other function name as well.

Reply:
Hi Saif, yes renamed all the applicable functions to avoid confusion.